tests/Homely.AspNetCore.Mvc.Helpers.Tests/TestControllerTests/ErrorTests.cs

```diff
@@ -0,0 +1,139 @@
+
+            var errors = new
+            {
+                errors = new List<ApiError>
+                {
+                    new ApiError
+                    {
+                        Message = "I'm a little tea pot, short and stout."
+                    }
+                }
+            };
+
+            // Act.
+            var response = await client.GetAsync("/test/error");
+
+            // Assert.
+            response.StatusCode.ShouldBe(HttpStatusCode.UpgradeRequired);
+            await response.Content.ShouldLookLike(errors);
+
+            //// We can't use the normal model<-->model compare because the response contains
+            //// a massive stack trace now (by design) and that's too hard to setup. So we'll
+            //// do a simple property-by-property check.
+            //var apiErrorsJson = await response.Content.ReadAsStringAsync();
+            //var apiErrorResult = JsonConvert.DeserializeObject<ApiErrorResult>(apiErrorsJson);
+            //apiErrorResult.Errors.ShouldLookLike(errors);
+            //apiErrorResult.StackTrace.ShouldBeNullOrWhiteSpace();
+        }
+
+        [Fact]
+        public async Task GivenAGetRequestAndAIncludeStackTrace_Error_ReturnsAnHttp500()
+        {
+            // Arrange.
+
+            // We want to use our own startup stuff so we can define our error callback.
+            var server = new TestServer(new WebHostBuilder().UseStartup<TestStartupWithCustomJsonExceptionPageIncludeStackTrace>());
+            var client = server.CreateClient();
+
+            // Act.
+            var response = await client.GetAsync("/test/error");
+
+            // Assert.
+            response.StatusCode.ShouldBe(HttpStatusCode.InternalServerError);
+
+            // We can't use the normal model<-->model compare because the response contains
+            // a massive stack trace now (by design) and that's too hard to setup. So we'll
+            // do a simple property-by-property check.
+            var apiErrorsJson = await response.Content.ReadAsStringAsync();
+            var apiErrorResult = JsonConvert.DeserializeObject<ApiErrorResult>(apiErrorsJson);
+            apiErrorResult.Errors.ShouldLookLike(DefaultErrorViewModel.Errors);
+            apiErrorResult.StackTrace.ShouldNotBeNullOrWhiteSpace();
+        }
+
+        [Fact]
+        public async Task GivenAnAjaxGetRequestWithCorsEnabled_Error_ReturnsAnHttp500()
+        {
+            // Arrange.
+
+            // We want to use our own startup stuff so we can define our error callback.
+            var server = new TestServer(new WebHostBuilder().UseStartup<TestStartupWithCors>());
+            var client = server.CreateClient();
+            client.DefaultRequestHeaders.Add("Origin", "http://foo.example");
+
+            // Act.
+            var response = await client.GetAsync("/test/error");
+
+            // Assert.
+            response.StatusCode.ShouldBe(HttpStatusCode.InternalServerError);

```

Comment:
should this be 500? not really a _server_ error?

Reply:
I don't understand what you are asking?

This test (`GivenAnAjaxGetRequestWithCorsEnabled_Error_ReturnsAnHttp500`) is a test to see that we can return the correct CORS headers, given the request is a CORS request _and_ an `InternalServerError` (HTTP 500 error) has occurred.

Reply:
Ah, i see now. I got confused. I thought you were testing here than when the CORS header was _rejected_, it should return a 500. (should be a 401/403). All good 😄